### ISSUE TITLE
update_cells no longer supports non-contiguous cells (overwrites unlisted cells within the range)

### ISSUE THREAD

**Issue by CianH** [burnash/gspread#538]
This appears to be a regression in the gpsread use of the v4 API (was previously working: see #177)

### Environment info
Operating System: Windows / Linux
Python version: 2.7.14
gspread version: 2.1.1

### Steps to reproduce
1. Make a list of cells that are non-contiguous, e.g. [<Cell R7C2 '2018'>, <Cell R20C2 '2004'>]
2. Call update_cells, passing the list
3. Cells R7C2 _through_ R20C2 inclusive get updated, with only R7C2 and R20C2 having values, R8-R19C2 all get zeroed out.

### Stack trace or other output that would be helpful
Compare update_cells implementation in `/v4/models.py` with `models.py`

```
>>> worksheet.range('B7:B20')
[<Cell R7C2 ''>, <Cell R8C2 ''>, <Cell R9C2 u'2013'>, <Cell R10C2 u'2013'>, <Cell R11C2 u'2015'>, <Cell R12C2 u'2015'>, <Cell R13C2 ''>, <Cell R14C2 u'2017'>, <Cell R15C2 u'2015'>, <Cell R16C2 u'2017'>, <Cell R17C2 u'2016'>, <Cell R18C2 u'2014'>, <Cell R19C2 u'2012'>, <Cell R20C2 ''>]
>>> cells = worksheet.range('B7:B20')
>>> other_cells = []
>>> other_cells.append(cells[0])
>>> other_cells.append(cells[len(cells)-1])
>>> other_cells
[<Cell R7C2 ''>, <Cell R20C2 ''>]
>>> other_cells[0].value = "test"
>>> other_cells[1].value = "testing"
>>> len(other_cells)
2
>>> worksheet.update_cells(other_cells)
{u'spreadsheetId': u'**removed**', u'updatedRange': u'Collection!B7:B20', u'updatedCells': 14, u'updatedRows': 14, u'updatedColumns': 1}
>>> worksheet.range('B7:B20')
[<Cell R7C2 u'test'>, <Cell R8C2 ''>, <Cell R9C2 ''>, <Cell R10C2 ''>, <Cell R11C2 ''>, <Cell R12C2 ''>, <Cell R13C2 ''>, <Cell R14C2 ''>, <Cell R15C2 ''>, <Cell R16C2 ''>, <Cell R17C2 ''>, <Cell R18C2 ''>, <Cell R19C2 ''>, <Cell R20C2 u'testing'>]
```

**Comment by ADraginda** [comment]
I wanted to confirm this issue. Non-continuous rows are fine (will only update those you change) but non-continuous columns will fill in with blanks. (From the first col to the last column updated - columns outside this range are untouched). 

I was working with gspread version `3.0.0`

**Comment by AlexRiina** [comment]
Based on https://developers.google.com/sheets/api/samples/writing it looks updating the [cell_list_to_rect](
https://github.com/burnash/gspread/blob/31bf1a13e1f6c149d4757f8cb195b3a763a37bdb/gspread/utils.py#L214)
to default to `None` instead of empty string will preserve the entries

> It is possible, when writing values to a range, to avoid changing some of the existing cells by setting the corresponding array elements to null. It is also possible to clear a cell by writing an empty string to it.

**Comment by ADraginda** [comment]
@CianH Using @AlexRiina 's suggestion, I verified that setting to None does work. I'll be submitting a PR, but in the interim you can workaround the issue by setting the `.value` for every cell to None before doing your actual value setting. If you do that, then only your changes will be applies.

**Comment by CianH** [comment]
@ADraginda nice, thanks for the workaround. For now, however, I've reverted to v0.6.2 since it's reliable for my use case.

**Comment by AlexRiina** [comment]
@Adraginda

Thanks! I wasn't able to get the test suite without some setup so I forked and made a version to use in the mean time.

I tried something like your proposed workaround but ran into some issues when i built a list of cells where the first cell wasn't the top and left most or the last cell want the bottom and right most.

Here's a patch of you run into the same thing https://github.com/AlexRiina/gspread/commit/6648c470c50691afe4b48b95a61f53bf59d70b87

**Comment by ADraginda** [comment]
Hm, I think you might have an issue with your work around. To test, I created a sheet where all of the values are set to 1

```
# starting with an open sheet
worksheet = sheet.get_worksheet(0)
cell_list = worksheet.range('A1:G10')
for cell in cell_list:
    cell.value = None
cell_list[20].value = 'XXXXX'
worksheet.update_cells(cell_array)
```

So I'm only writing one cell ... and this worked for me. No errors and it only updated the one cell. I haven't opened my PR yet, but it will contain this logic - which works for me separately:

```
def cell_list_to_rect(cell_list):
    if not cell_list:
        return []

    rows = defaultdict(lambda: {})

    row_offset = cell_list[0].row
    col_offset = cell_list[0].col

    for cell in cell_list:
        row = rows.setdefault(int(cell.row) - row_offset, {})
        row[cell.col - col_offset] = cell.value

    if not rows:
        return []

    all_row_keys = chain.from_iterable(row.keys() for row in rows.values())
    rect_cols = range(max(all_row_keys) + 1)
    rect_rows = range(max(rows.keys()) + 1)

    # Return the values of the cells as a list of lists where each sublist
    # contains all of the values for one row. The Google API requires a rectangle
    # of updates, so if a cell isn't present in the input cell_list, then the
    # value will be None and will not be updated.
    return [[rows[i].get(j) for j in rect_cols] for i in rect_rows]
```

**Comment by ADraginda** [comment]
@AlexRiina you were right - we also needed that min/max bit you posted there. Thanks. I added it to my PR.

**Comment by burnash** [comment]
@CianH thank you for reporting the issue. It was fixed in #553 and released in version 3.0.1 on pypi. Please upgrade the library and check.

@AlexRiina thank you very much for the hint and comments. @ADraginda thank you again for the pull request with the fix.

I'm closing this issue now.
